File: frontend/src/lib/types/skill_roadmap.ts
/**
 * Strongly typed TypeScript contracts for SkillForge AI: Static Skill Roadmap Rebuild.
 * Phase 5: Frontend API Client & TypeScript Contracts.
 *
 * Strictly matches backend Pydantic schemas in `backend/app/schemas/skill_roadmap.py`
 * and REST endpoints in `backend/app/api/v1/roadmaps.py`.
 *
 * Decoupled from P4 Personalized Career Roadmap contracts (`roadmap.ts`).
 */

// -----------------------------------------------------------------------------
// Core Literal & Enum Types
// -----------------------------------------------------------------------------

/**
 * Valid static roadmap skill learning statuses.
 * Matches backend UserRoadmapProgress and RoadmapSkillItem.user_status.
 */
export type RoadmapSkillStatus = "NOT_STARTED" | "LEARNING" | "DONE" | "SKIPPED";

/**
 * Valid practice problem completion statuses.
 * Matches backend UserPracticeProgress and PracticeProblemItem.user_status.
 */
export type PracticeProblemStatus = "NOT_STARTED" | "IN_PROGRESS" | "COMPLETED";

/**
 * Static roadmap view ordering mode:
 * - curated: canonical stage-by-stage pedagogy
 * - recommended: deterministic Kahn topological sort obeying prerequisite DAG
 */
export type RoadmapOrdering = "curated" | "recommended";

/**
 * Resource media types supported by static roadmap learning resources.
 */
export type LearningResourceType = "DOCUMENTATION" | "YOUTUBE";

/**
 * Difficulty tiers for practice problems.
 */
export type PracticeProblemDifficulty = "BEGINNER" | "INTERMEDIATE" | "ADVANCED";

// -----------------------------------------------------------------------------
// Learning Resource Schemas
// -----------------------------------------------------------------------------

export interface LearningResourceItem {
  id: string;
  resource_type: LearningResourceType | string;
  title: string;
  url: string;
  description: string;
}

// -----------------------------------------------------------------------------
// Prerequisite Schemas
// -----------------------------------------------------------------------------

export interface RoadmapPrerequisiteItem {
  skill_id: string;
  skill_name: string;
  skill_slug: string;
  difficulty: string;
}

export type StaticRoadmapPrerequisiteItem = RoadmapPrerequisiteItem;

// -----------------------------------------------------------------------------
// Practice Problem Schemas
// -----------------------------------------------------------------------------

export interface PracticeProblemItem {
  problem_id: string;
  title: string;
  difficulty: PracticeProblemDifficulty | string;
  order: number;
  objective: string;
  problem_statement: string;
  requirements: string[];
  concepts_tested: string[];
  expected_outcome: string;
  optional_hints: string[];
  user_status?: PracticeProblemStatus | string;

  // Optional convenience aliases for UI consumers
  description?: string;
  hints?: string[];
  starter_code?: string | null;
  test_cases?: unknown[];
}

// -----------------------------------------------------------------------------
// Roadmap Skill Schemas
// -----------------------------------------------------------------------------

export interface RoadmapSkillItem {
  id: string;
  stage_id: string;
  roadmap_id: string;
  canonical_skill_id?: string | null;
  name: string;
  slug: string;
  description: string;
  difficulty: string;
  skill_order: number;
  key_topics: string[];
  practice_project?: string | null;
  practice_problems: PracticeProblemItem[];
  role_relevance?: string | null;
  user_status?: RoadmapSkillStatus | string;
  prerequisites: RoadmapPrerequisiteItem[];
  resources: LearningResourceItem[];
}

// -----------------------------------------------------------------------------
// Roadmap Stage Schemas
// -----------------------------------------------------------------------------

export interface RoadmapStageItem {
  id: string;
  roadmap_id: string;
  name: string;
  description?: string | null;
  stage_order: number;
  total_skills: number;
  completed_skills: number;
  skills: RoadmapSkillItem[];
}

// -----------------------------------------------------------------------------
// Roadmap Catalog List Schemas
// -----------------------------------------------------------------------------

export interface RoadmapListItem {
  id: string;
  role_id?: string | null;
  slug: string;
  title: string;
  domain: string;
  category: string;
  description: string;
  version: string;
  has_market_data: boolean;
  total_stages: number;
  total_skills: number;
  last_reviewed: string;
}

export interface RoadmapListResponse {
  data: RoadmapListItem[];
  total: number;
}

// -----------------------------------------------------------------------------
// Roadmap Detail & Summary Schemas
// -----------------------------------------------------------------------------

export interface RoadmapSummary {
  total_skills: number;
  completed_skills: number;
  learning_skills: number;
  skipped_skills: number;
  not_started_skills: number;
  progress_percentage: number;
}

export interface RoadmapDetailData {
  roadmap: RoadmapListItem;
  ordering: RoadmapOrdering | string;
  summary: RoadmapSummary;
  stages: RoadmapStageItem[];
  recommended_skills?: RoadmapSkillItem[] | null;
}

export interface RoadmapDetailResponse {
  data: RoadmapDetailData;
}

// -----------------------------------------------------------------------------
// User Progress Update & Response Schemas
// -----------------------------------------------------------------------------

export interface UserProgressUpdateRequest {
  status: RoadmapSkillStatus;
}

export interface UserProgressItem {
  id: string;
  user_id: string;
  roadmap_skill_id: string;
  status: RoadmapSkillStatus | string;
  created_at: string;
  updated_at: string;
}

export interface UserProgressResponse {
  success: boolean;
  data: UserProgressItem;
}

// -----------------------------------------------------------------------------
// User Practice Progress Update & Response Schemas
// -----------------------------------------------------------------------------

export interface UserPracticeProgressUpdateRequest {
  status: PracticeProblemStatus;
}

export interface UserPracticeProgressItem {
  id: string;
  user_id: string;
  roadmap_skill_id: string;
  problem_id: string;
  status: PracticeProblemStatus | string;
  completed_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface UserPracticeProgressResponse {
  success: boolean;
  data: UserPracticeProgressItem;
}

// -----------------------------------------------------------------------------
// User Roadmap Progress Summary & Map Read Models
// -----------------------------------------------------------------------------

export interface UserRoadmapProgressSummary {
  roadmap_id: string;
  total_skills: number;
  completed_skills: number;
  learning_skills: number;
  skipped_skills: number;
  not_started_skills: number;
  progress_percentage: number;
  total_practice_problems: number;
  completed_practice_problems: number;
}

/**
 * Key-value mapping of roadmap_skill_id to current status.
 * Returned by GET /api/v1/users/me/roadmap-progress when summary=false.
 */
export type UserRoadmapProgressMap = Record<string, RoadmapSkillStatus | string>;

File: backend/app/schemas/skill_roadmap.py
"""
Pydantic Schemas for Static Skill Roadmap Catalog and User Progress.
Phase 3: Static Roadmap Schemas & Service.
"""

from datetime import datetime
from typing import List, Literal, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field, field_validator


# -----------------------------------------------------------------------------
# Learning Resource Schemas
# -----------------------------------------------------------------------------

class LearningResourceItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    resource_type: str = Field(..., description="Resource type: DOCUMENTATION or YOUTUBE")
    title: str = Field(..., description="Resource title")
    url: str = Field(..., description="Resource URL")
    description: str = Field(..., description="Resource description")


# -----------------------------------------------------------------------------
# Prerequisite Schemas
# -----------------------------------------------------------------------------

class RoadmapPrerequisiteItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    skill_id: UUID
    skill_name: str
    skill_slug: str
    difficulty: str


# -----------------------------------------------------------------------------
# Practice Problem Schemas
# -----------------------------------------------------------------------------

class PracticeProblemItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    problem_id: str
    title: str
    difficulty: str = Field(default="BEGINNER", description="BEGINNER, INTERMEDIATE, or ADVANCED")
    order: int
    objective: str
    problem_statement: str
    requirements: List[str] = Field(default_factory=list)
    concepts_tested: List[str] = Field(default_factory=list)
    expected_outcome: str
    optional_hints: List[str] = Field(default_factory=list)
    user_status: str = Field(
        default="NOT_STARTED",
        description="User practice status: NOT_STARTED, IN_PROGRESS, COMPLETED",
    )


# -----------------------------------------------------------------------------
# Roadmap Skill Schemas
# -----------------------------------------------------------------------------

class RoadmapSkillItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    stage_id: UUID
    roadmap_id: UUID
    canonical_skill_id: Optional[UUID] = None
    name: str
    slug: str
    description: str
    difficulty: str
    skill_order: int
    key_topics: List[str] = Field(default_factory=list)
    practice_project: Optional[str] = None
    practice_problems: List[PracticeProblemItem] = Field(default_factory=list)
    role_relevance: Optional[str] = None
    user_status: str = Field(
        default="NOT_STARTED",
        description="User learning status: NOT_STARTED, LEARNING, DONE, SKIPPED",
    )
    prerequisites: List[RoadmapPrerequisiteItem] = Field(default_factory=list)
    resources: List[LearningResourceItem] = Field(default_factory=list)


# -----------------------------------------------------------------------------
# Roadmap Stage Schemas
# -----------------------------------------------------------------------------

class RoadmapStageItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    roadmap_id: UUID
    name: str
    description: Optional[str] = None
    stage_order: int
    total_skills: int
    completed_skills: int
    skills: List[RoadmapSkillItem] = Field(default_factory=list)


# -----------------------------------------------------------------------------
# Roadmap Catalog List Schemas
# -----------------------------------------------------------------------------

class RoadmapListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    role_id: Optional[UUID] = None
    slug: str
    title: str
    domain: str
    category: str
    description: str
    version: str
    has_market_data: bool
    total_stages: int
    total_skills: int
    last_reviewed: datetime


class RoadmapListResponse(BaseModel):
    data: List[RoadmapListItem]
    total: int


# -----------------------------------------------------------------------------
# Roadmap Detail & Summary Schemas
# -----------------------------------------------------------------------------

class RoadmapSummary(BaseModel):
    total_skills: int
    completed_skills: int
    learning_skills: int
    skipped_skills: int
    not_started_skills: int
    progress_percentage: float


class RoadmapDetailData(BaseModel):
    roadmap: RoadmapListItem
    ordering: str = Field(default="curated", description="Viewing mode: curated or recommended")
    summary: RoadmapSummary
    stages: List[RoadmapStageItem]
    recommended_skills: Optional[List[RoadmapSkillItem]] = None


class RoadmapDetailResponse(BaseModel):
    data: RoadmapDetailData


# -----------------------------------------------------------------------------
# User Progress Update & Response Schemas
# -----------------------------------------------------------------------------

class UserProgressUpdateRequest(BaseModel):
    status: str = Field(..., description="Updated status: NOT_STARTED, LEARNING, DONE, SKIPPED")

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str) -> str:
        if not isinstance(v, str):
            raise ValueError("Status must be a string.")
        norm = v.strip().upper()
        allowed = {"NOT_STARTED", "LEARNING", "DONE", "SKIPPED"}
        if norm not in allowed:
            raise ValueError(f"Invalid status '{v}'. Allowed statuses: {', '.join(sorted(allowed))}")
        return norm


class UserProgressItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    roadmap_skill_id: UUID
    status: str
    created_at: datetime
    updated_at: datetime


class UserProgressResponse(BaseModel):
    success: bool = True
    data: UserProgressItem


# -----------------------------------------------------------------------------
# User Practice Progress Update & Response Schemas
# -----------------------------------------------------------------------------

class UserPracticeProgressUpdateRequest(BaseModel):
    status: str = Field(..., description="Updated status: NOT_STARTED, IN_PROGRESS, COMPLETED")

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str) -> str:
        if not isinstance(v, str):
            raise ValueError("Status must be a string.")
        norm = v.strip().upper()
        allowed = {"NOT_STARTED", "IN_PROGRESS", "COMPLETED"}
        if norm not in allowed:
            raise ValueError(f"Invalid status '{v}'. Allowed statuses: {', '.join(sorted(allowed))}")
        return norm


class UserPracticeProgressItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    roadmap_skill_id: UUID
    problem_id: str
    status: str
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class UserPracticeProgressResponse(BaseModel):
    success: bool = True
    data: UserPracticeProgressItem


# -----------------------------------------------------------------------------
# User Roadmap Progress Summary Read Model
# -----------------------------------------------------------------------------

class UserRoadmapProgressSummary(BaseModel):
    roadmap_id: UUID
    total_skills: int
    completed_skills: int
    learning_skills: int
    skipped_skills: int
    not_started_skills: int
    progress_percentage: float
    total_practice_problems: int
    completed_practice_problems: int
